validation/coins: limit MemPoolAccept access to coins cache + make it responsible for uncaching

glozow opened the pull request:
This PR is a subpart of #20833 which enables package validation through `testmempoolaccept`.
The first commit addresses some leftover comments from #21062.


When validating unconfirmed transactions, we want to account for all coins brought in from disk so we can uncache them if the tx is invalid. The disconnection of `CCoinsView`s right now doesn't quite achieve what's it's trying to do (prevent `MemPoolAccept` from unnecessarily accessing the coins cache and forgetting to uncache later). This PR improves it by doing 2 things:
(1) Disconnect `m_viewmempool` from the coins cache when we don't need it to be connected.
(2) Make `MemPoolAccept` destructor clean up after itself by uncaching coins accessed, instead of leaving it up to the caller. Only leave the coins cached when a transaction is accepted to mempool. In the future, if someone edits the `MemPoolAccept` class and forgets to consider uncaching, the default behavior is to uncache (which is safer).

glozow committed:
[validation] make MemPoolAccept responsible for uncaching coins

MemPoolAccept should clean up after itself (woohoo RAII), uncache all
coins by default, and explicitly decide to keep them on specific
occasions (i.e. when tx is submitted to mempool). This is safer and
easier to reason about than leaving it up to the caller.
Also allows us to make ATMPArgs always const.

Note: with this commit, we always uncache coins after a test_accept
regardless of validation results. Previously, we uncached whenever a tx
failed validation.

glozow committed:
doc/style followups in MempoolAcceptResult

glozow commented on src/validation.cpp:
```diff
@@ -671,7 +691,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     // we have all inputs cached now, so switch back to dummy (to protect
     // against bugs where we pull more inputs from disk that miss being added
     // to coins_to_uncache)
-    m_view.SetBackend(m_dummy);
+    m_viewmempool.SetBackend(m_dummy);
```
☝️ this is the main change 👆

jnewbery replied:
You switch `m_view.SetBackend(m_dummy)` to `m_viewmempool.SetBackend(m_dummy)` in commit _[validation] disconnect m_viewmempool from coins cache when unnecessary_.

I think perhaps in that commit you should just add `m_viewmempool.SetBackend(m_dummy)`, and then in the subsequent commit (_[validation] keep m_view->m_viewmempool connected_), you should remove `m_view.SetBackend(m_dummy)`. That keeps the commits more separated (the first is adding the m_viewmempool // CoinsTip() split, and the second is removing the m_view // m_viewmempool split).

glozow replied:
Good point! Done in the last push

ariard commented:
Do you think you can take the following diff in the doc followup commit ?

```
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
index ac4240447..f8f4a6d94 100644
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -896,7 +896,7 @@ static RPCHelpMan testmempoolaccept()
                         {
                             {RPCResult::Type::STR_HEX, "txid", "The transaction hash in hex"},
                             {RPCResult::Type::STR_HEX, "wtxid", "The transaction witness hash in hex"},
-                            {RPCResult::Type::BOOL, "allowed", "If the mempool allows this tx to be inserted"},
+                            {RPCResult::Type::BOOL, "allowed", "If the mempool and client specified maxfeerate allows this tx to be inserted"},
                             {RPCResult::Type::NUM, "vsize", "Virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted (only present when 'allowed' is true)"},
                             {RPCResult::Type::OBJ, "fees", "Transaction fees (only present if 'allowed' is true)",
                             {
```

That would close my [#21704](https://github.com/bitcoin/bitcoin/issues/21074) concerns for now :) ?

glozow committed:
[validation] disconnect m_viewmempool from coins cache when unnecessary

The hierarchy of CoinsViews here is:
m_view -> m_viewmempool -> chainstate coins cache

To prevent DoS, we want to limit access to coins cache and ensure we
uncache coins if the transaction turns out to be invalid.  When
m_viewmempool is always connected to the coins cache, it's harder reason
about this. We could accidentally access the coins cache if we use
m_viewmempool for something.

glozow commented:
Rebased

jnewbery reviewed (commented):
Final code looks good. I have one comment inline about the commit structure.

There's a small change in the final commit that you haven't communicated: currently, if a testmempoolaccept is successful, we keep any fetched coins in the cache. After this commit we don't. I think that's a fine change - if the transaction hasn't been added to the mempool, then there's no reason for the coins it spends to be added to the cache. The change is perhaps worth pointing out in the commit log though.

glozow commented:
Thanks for the review @jnewbery! Addressed your comments. Added some sentences about the uncaching behavior to https://github.com/bitcoin/bitcoin/commit/acef823e895b80fd7f4c2f5a44c3470fe46c1a6a commit log.

achow101 commented:
ACK acef823e895b80fd7f4c2f5a44c3470fe46c1a6a

ariard reviewed (commented):
Should you extend 34ab1ad7 commit message to explain why it's safe to access output in the mempool ? I would say for the mempool, we don't have overrun concerns, all the state is already in-memory. No size extension from disk is possible.

glozow commented on src/validation.cpp:
```diff
@@ -1088,20 +1104,10 @@ static MempoolAcceptResult AcceptToMemoryPoolWithTime(const CChainParams& chainp
 
-        for (const COutPoint& hashTx : coins_to_uncache)
-            active_chainstate.CoinsTip().Uncache(hashTx);
-    }
```
> After this change, it's spread in two locations : in case of success evaluation in AcceptSingleTransaction, in case of failure in MemPoolAccept destructor. 

Not really. The only location a reviewer has to check is wherever they _want_ to keep the coins in cache. Also, the change here means that if someone forgets to consider uncaching, it defaults to uncache all. This is safer.


> Also, I don't know about relying on RAII semantic, the destructor will be only called when object is going out of scope. In that case, does it mean we'll only cleanup when caller ends up ? For e.g, if the caller was BroadcastTransaction, may the coin cache cleanup on cs_main lock tacking before RelayTransaction ?

The two possible callers of `MemPoolAccept` are ATMP and ProcessNewPackage They invoke it like this:

```c
const MempoolAcceptResult result = MemPoolAccept(pool, active_chainstate).AcceptSingleTransaction(tx, args);
```

afaik it goes out of scope immediately, before the lock is released.

ariard replied:
> Not really. The only location a reviewer has to check is wherever they want to keep the coins in cache. 

Depends what you're willingly to convince yourself by reviewing. Assuming uncaching  in case of failures is the correct behavior, looking at the proposed changes, you have to a) verify that uncaching code is well-implemented in `MemPoolAccept` destructor, b) identify that caching conservation is covering all success paths and thus dissociate success code path from errors one in `AcceptSingleTransaction` and c) ensure the RAAI semantic is enforced as intended. IMHO, more complex than current one-place location.

That said, I think the code is correct, and each one review process and code style flavors might turn as really abstract discussion so not going to argue further :)

jnewbery replied:
> The two possible callers of MemPoolAccept are ATMP and ProcessNewPackage They invoke it like this:
> 
> `const MempoolAcceptResult result = MemPoolAccept(pool, active_chainstate).AcceptSingleTransaction(tx, args);`
> 
> afaik it goes out of scope immediately, before the lock is released.

I believe you're right. In that expression, `MemPoolAccept()` is a temporary, and the lifetime rules for temporaries are

> All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created, and if multiple temporary objects were created, they are destroyed in the order opposite to the order of creation. This is true even if that evaluation ends in throwing an exception.

https://en.cppreference.com/w/cpp/language/lifetime#Temporary_object_lifetime

Note that there are two exceptions to that rule, neither of which apply here:

> - The lifetime of a temporary object may be extended by binding to a const lvalue reference or to an rvalue reference (since C++11), see reference initialization for details.
> - The lifetime of a temporary object created when evaluating the default arguments of a default constructor used to initialize an element of an array ends before the next element of the array begins initialization.

jnewbery commented on src/validation.cpp:
```diff
@@ -1088,20 +1104,10 @@ static MempoolAcceptResult AcceptToMemoryPoolWithTime(const CChainParams& chainp
-    std::vector<COutPoint> coins_to_uncache;
-    MemPoolAccept::ATMPArgs args { chainparams, nAcceptTime, bypass_limits, coins_to_uncache, test_accept };
-
+    MemPoolAccept::ATMPArgs args { chainparams, nAcceptTime, bypass_limits, test_accept };
```
You can const this if you want

glozow replied:
I always wanna const

jnewbery commented on src/validation.cpp:
```diff
@@ -1088,20 +1104,10 @@ static MempoolAcceptResult AcceptToMemoryPoolWithTime(const CChainParams& chainp
-        for (const COutPoint& hashTx : coins_to_uncache)
-            active_chainstate.CoinsTip().Uncache(hashTx);
-    }
     // After we've (potentially) uncached entries, ensure our coins cache is still within its size limits
```
Consider removing "After we've (potentially) uncached entries". That was referring to the line above. Now that it's removed, that part of the comment is not particularly relevant here.

jnewbery reviewed (commented):
utACK acef823e895b80fd7f4c2f5a44c3470fe46c1a6a

glozow commented:
I'm working on #20833 and realized I don't need to keep `m_view` and `m_viewmempool` connected for it to work 😛 so I took out the 3rd commit (_[validation] keep m_view->m_viewmempool connected_). @ariard you might find this PR a bit more acceptable now :). I'll try to add https://github.com/bitcoin/bitcoin/pull/21146#issuecomment-778213317 to a different PR where I'm touching the RPC code.

Also addressed @jnewbery's comments https://github.com/bitcoin/bitcoin/pull/21146#discussion_r582119695 and https://github.com/bitcoin/bitcoin/pull/21146#discussion_r582120303.

glozow commented:
I still think this would be a nice change but since I don't need this for #20833 anymore, I'm closing this to focus on that instead.
